core/state/journal.go

```diff
@@ -163,10 +167,16 @@ func (ch resetObjectChange) revert(s *StateDB) {
 		delete(s.stateObjectsDestruct, ch.prev.address)
 	}
 	if ch.prevAccount != nil {
-		s.snapAccounts[ch.prev.addrHash] = ch.prevAccount
+		s.accounts[ch.prev.addrHash] = ch.prevAccount
 	}
 	if ch.prevStorage != nil {
-		s.snapStorage[ch.prev.addrHash] = ch.prevStorage
+		s.storages[ch.prev.addrHash] = ch.prevStorage
+	}
+	if ch.prevAccountOriginExist {
+		s.accountsOrigin[ch.prev.addrHash] = ch.prevAccountOrigin
+	}
```

Comment:
We should document, somewhere, what this means. For example, if `prevAccountOriginExist` is `true`, can `prevAccountOrigin` be `nil`? And what kind of scenario would lead to that? 
Simiarly, can `prevAccountOriginExist` be `false` and `prevAccountOrigin` be non-`nil`? If so, how? 

And if they're both 'strongly' tied to eachother, do we really need `prevAccountOriginExist`?

Reply:
It's possible that `prevAccountOriginExist` is true and `prevAccountOrigin` is nil.

`nil` means the account was not present and it's created in this block. If we want
to reset this block, we need the explicit flag to distinguish:
- accountOrigin is nil
- accountOrigin is not existent

Regarding `prevAccountOriginExist ` is false while `prevAccountOrigin` is non-nil, it's an impossible combo.